File: sdk/sdk-go/sdk/ipc.go
package sdk

// Minimal client for the Glimpse IPC socket.
//
// IPC(service) resolves a *Subscriber (no I/O — the connection is opened
// lazily). Subscriber.Listen subscribes to an event channel and streams
// decoded Events over a channel; Subscriber.Dispatch sends an action and
// awaits the server ack on a one-shot connection. The wire protocol matches
// the `glimpse-shell watch` / `dispatch` CLIs.

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
)

// Event is one decoded event line; Fields values are unescaped.
type Event struct {
	Name   string
	Ts     int64
	Fields map[string]string
}

// Subscriber is a resolved IPC endpoint. Cheap to create; holds only the
// socket path (or the error from resolving it, surfaced on first use).
type Subscriber struct {
	socket     string
	resolveErr error

	mu  sync.Mutex
	err error
}

// IPC resolves the Subscriber for service (use "shell", or "" for it, to
// reach the panel). The socket is <dir>/<service>.sock — "shell" maps to
// ipc.sock — where <dir> is $GLIMPSE_IPC_DIR, else $XDG_RUNTIME_DIR/glimpse.
// No connection is made here; resolution errors surface from Listen/Dispatch.
func IPC(service string) *Subscriber {
	socket, err := socketPath(service)
	return &Subscriber{socket: socket, resolveErr: err}
}

func socketPath(service string) (string, error) {
	if service == "" {
		service = "shell"
	}
	var dir string
	if v := os.Getenv("GLIMPSE_IPC_DIR"); v != "" {
		dir = v
	} else if x := os.Getenv("XDG_RUNTIME_DIR"); x != "" {
		dir = filepath.Join(x, "glimpse")
	} else {
		return "", fmt.Errorf(
			"neither GLIMPSE_IPC_DIR nor XDG_RUNTIME_DIR is set; " +
				"cannot locate the Glimpse IPC socket")
	}
	name := "ipc.sock"
	if service != "shell" {
		name = service + ".sock"
	}
	return filepath.Join(dir, name), nil
}

// Err returns why a Listen stream ended (nil for a clean EOF or a canceled
// context). Read it after the Listen channel closes.
func (s *Subscriber) Err() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.err
}

func (s *Subscriber) setErr(err error) {
	s.mu.Lock()
	s.err = err
	s.mu.Unlock()
}

// Listen subscribes to channel (an exact name, a "prefix.*" pattern, or "*")
// and streams events until the server closes the connection or ctx is
// canceled. The returned channel is closed when the stream ends; inspect
// Err afterwards for the terminal cause.
//
// The caller MUST cancel ctx when done (e.g. break out of the range loop and
// cancel): a reader goroutine and the socket fd stay alive until the server
// closes the connection or ctx is canceled, so abandoning the channel
// without canceling leaks both.
func (s *Subscriber) Listen(ctx context.Context, channel string) (<-chan Event, error) {
	conn, br, err := s.connect(ctx)
	if err != nil {
		return nil, err
	}
	if _, err := io.WriteString(conn, "subscribe "+channel+"\n"); err != nil {
		conn.Close()
		return nil, err
	}
	ch := make(chan Event)
	stop := context.AfterFunc(ctx, func() { conn.Close() })
	go func() {
		defer close(ch)
		defer stop()
		defer conn.Close()
		for {
			line, err := br.ReadString('\n')
			if err != nil {
				if ctx.Err() == nil && err != io.EOF {
					s.setErr(err)
				}
				return
			}
			line = strings.TrimSpace(line)
			if line == "" {
				continue
			}
			select {
			case ch <- parseEvent(line):
			case <-ctx.Done():
				return
			}
		}
	}()
	return ch, nil
}

// Dispatch sends action with params on a fresh connection and awaits the
// ack. It returns the extra ack fields, or an error if connecting fails or
// the server replies ok=false.
func (s *Subscriber) Dispatch(
	ctx context.Context, action string, params map[string]string,
) (map[string]string, error) {
	if err := validateToken("action", action, false); err != nil {
		return nil, err
	}
	keys := make([]string, 0, len(params))
	for k := range params {
		if err := validateToken("param key", k, true); err != nil {
			return nil, err
		}
		keys = append(keys, k)
	}
	// Deterministic wire order (Go map iteration is randomized).
	sort.Strings(keys)

	conn, br, err := s.connect(ctx)
	if err != nil {
		return nil, err
	}
	defer conn.Close()
	stop := context.AfterFunc(ctx, func() { conn.Close() })
	defer stop()

	var b strings.Builder
	b.WriteString(action)
	for _, k := range keys {
		b.WriteByte(' ')
		b.WriteString(k)
		b.WriteByte('=')
		b.WriteString(escapeValue(params[k]))
	}
	b.WriteByte('\n')
	if _, err := io.WriteString(conn, b.String()); err != nil {
		return nil, err
	}
	line, err := br.ReadString('\n')
	if err != nil {
		return nil, fmt.Errorf("reading IPC ack: %w", err)
	}
	return parseAck(strings.TrimSpace(line))
}

func (s *Subscriber) connect(ctx context.Context) (net.Conn, *bufio.Reader, error) {
	if s.resolveErr != nil {
		return nil, nil, s.resolveErr
	}
	var d net.Dialer
	conn, err := d.DialContext(ctx, "unix", s.socket)
	if err != nil {
		return nil, nil, fmt.Errorf(
			"cannot connect to IPC socket at %s: %w", s.socket, err)
	}
	br := bufio.NewReader(conn)
	hello, err := br.ReadString('\n')
	if err != nil {
		conn.Close()
		return nil, nil, fmt.Errorf("IPC server closed connection before hello: %w", err)
	}
	if !strings.HasPrefix(hello, "hello") {
		conn.Close()
		return nil, nil, fmt.Errorf(
			"unexpected IPC greeting: %s", strings.TrimSpace(hello))
	}
	return conn, br, nil
}

// validateToken rejects action names / field keys that would forge extra
// tokens or whole client lines: the wire protocol splits on whitespace and
// never unescapes the command name or a key (only values are escaped).
func validateToken(label, token string, forbidEq bool) error {
	if token == "" {
		return fmt.Errorf("IPC %s must not be empty", label)
	}
	if strings.ContainsAny(token, " \t\n\r\f\v") {
		return fmt.Errorf("IPC %s %q must not contain whitespace", label, token)
	}
	if forbidEq && strings.Contains(token, "=") {
		return fmt.Errorf("IPC param key %q must not contain '='", token)
	}
	return nil
}

func escapeValue(s string) string {
	r := strings.NewReplacer(
		"\\", "\\\\",
		"\n", "\\n",
		"\t", "\\t",
		" ", "\\s",
	)
	return r.Replace(s)
}

func unescapeValue(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	for i := 0; i < len(s); i++ {
		if s[i] != '\\' {
			b.WriteByte(s[i])
			continue
		}
		i++
		if i >= len(s) {
			b.WriteByte('\\')
			break
		}
		switch s[i] {
		case 's':
			b.WriteByte(' ')
		case 'n':
			b.WriteByte('\n')
		case 't':
			b.WriteByte('\t')
		case '\\':
			b.WriteByte('\\')
		default:
			b.WriteByte('\\')
			b.WriteByte(s[i])
		}
	}
	return b.String()
}

func parseEvent(line string) Event {
	tokens := strings.Fields(line)
	ev := Event{Fields: map[string]string{}}
	if len(tokens) > 0 {
		ev.Name = tokens[0]
	}
	for _, token := range tokens[1:] {
		key, raw, ok := strings.Cut(token, "=")
		if !ok {
			continue
		}
		value := unescapeValue(raw)
		if key == "ts" {
			if n, err := strconv.ParseInt(value, 10, 64); err == nil {
				ev.Ts = n
				continue
			}
		}
		ev.Fields[key] = value
	}
	return ev
}

func parseAck(line string) (map[string]string, error) {
	tokens := strings.Fields(line)
	if len(tokens) == 0 || tokens[0] != "ack" {
		return nil, fmt.Errorf("expected an ack, got: %s", line)
	}
	ok := false
	fields := map[string]string{}
	for _, token := range tokens[1:] {
		key, raw, has := strings.Cut(token, "=")
		if !has {
			continue
		}
		value := unescapeValue(raw)
		if key == "ok" {
			ok = value == "true"
		} else {
			fields[key] = value
		}
	}
	if !ok {
		if msg, has := fields["error"]; has {
			return nil, fmt.Errorf("%s", msg)
		}
		return nil, fmt.Errorf("command failed")
	}
	return fields, nil
}

File: sdk/sdk-rs/src/ipc.rs
//! Minimal client for the Glimpse IPC socket.
//!
//! [`ipc`] resolves a [`Subscriber`] for a service (no I/O — the connection
//! is opened lazily). [`Subscriber::listen`] subscribes to an event channel
//! and streams decoded [`Event`]s; [`Subscriber::dispatch`] sends an action
//! and awaits the server ack on a one-shot connection. The wire protocol is
//! the same line format the `glimpse-shell watch` / `dispatch` CLIs speak.

use std::collections::HashMap;
use std::env;
use std::path::PathBuf;

use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader, Lines};
use tokio::net::UnixStream;
use tokio::net::unix::{OwnedReadHalf, OwnedWriteHalf};

use crate::AppletResult;

/// One decoded event line: `name key=value … ts=<epoch>` with values
/// unescaped.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Event {
    pub name: String,
    pub ts: u64,
    pub fields: HashMap<String, String>,
}

/// A resolved IPC endpoint. Cheap to create; holds only the socket path.
#[derive(Debug, Clone)]
pub struct Subscriber {
    socket: PathBuf,
}

/// Resolve the [`Subscriber`] for `service` (use `"shell"` for the panel).
///
/// The socket is `<dir>/<service>.sock` — `shell` maps to `ipc.sock` — where
/// `<dir>` is `$GLIMPSE_IPC_DIR`, else `$XDG_RUNTIME_DIR/glimpse`. No
/// connection is made here; it happens in [`Subscriber::listen`] /
/// [`Subscriber::dispatch`].
pub fn ipc(service: &str) -> AppletResult<Subscriber> {
    Ok(Subscriber {
        socket: socket_path(service)?,
    })
}

fn socket_path(service: &str) -> AppletResult<PathBuf> {
    let service = if service.is_empty() { "shell" } else { service };
    let dir = if let Some(d) = env::var_os("GLIMPSE_IPC_DIR") {
        PathBuf::from(d)
    } else if let Some(x) = env::var_os("XDG_RUNTIME_DIR") {
        PathBuf::from(x).join("glimpse")
    } else {
        return Err("neither GLIMPSE_IPC_DIR nor XDG_RUNTIME_DIR is set; \
                    cannot locate the Glimpse IPC socket"
            .into());
    };
    let file = if service == "shell" {
        "ipc.sock".to_owned()
    } else {
        format!("{service}.sock")
    };
    Ok(dir.join(file))
}

impl Subscriber {
    /// Subscribe to `channel` (an exact name, a `prefix.*` pattern, or `*`)
    /// and stream events until the server closes the connection.
    pub async fn listen(&self, channel: &str) -> AppletResult<EventStream> {
        let mut conn = self.connect().await?;
        conn.writer
            .write_all(format!("subscribe {channel}\n").as_bytes())
            .await?;
        Ok(EventStream { lines: conn.lines })
    }

    /// Dispatch `action` with `params` on a fresh connection and await the
    /// ack. Returns the extra ack fields on success; errors if the server
    /// replies `ok=false`.
    pub async fn dispatch<I, K, V>(
        &self,
        action: &str,
        params: I,
    ) -> AppletResult<HashMap<String, String>>
    where
        I: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: AsRef<str>,
    {
        validate_token("action", action, false)?;
        let mut pairs: Vec<(String, String)> = Vec::new();
        for (k, v) in params {
            validate_token("param key", k.as_ref(), true)?;
            pairs.push((k.as_ref().to_owned(), v.as_ref().to_owned()));
        }
        let mut conn = self.connect().await?;
        let mut line = String::from(action);
        for (k, v) in &pairs {
            line.push(' ');
            line.push_str(k);
            line.push('=');
            line.push_str(&escape(v));
        }
        line.push('\n');
        conn.writer.write_all(line.as_bytes()).await?;
        let ack = conn
            .lines
            .next_line()
            .await?
            .ok_or("IPC server closed connection without ack")?;
        parse_ack(&ack)
    }

    async fn connect(&self) -> AppletResult<Conn> {
        let stream = UnixStream::connect(&self.socket).await.map_err(|e| {
            format!(
                "cannot connect to IPC socket at {}: {e}",
                self.socket.display()
            )
        })?;
        let (reader, writer) = stream.into_split();
        let mut lines = BufReader::new(reader).lines();
        match lines.next_line().await? {
            Some(l) if l.starts_with("hello") => {}
            Some(l) => return Err(format!("unexpected IPC greeting: {l}").into()),
            None => return Err("IPC server closed connection before hello".into()),
        }
        Ok(Conn { writer, lines })
    }
}

struct Conn {
    writer: OwnedWriteHalf,
    lines: Lines<BufReader<OwnedReadHalf>>,
}

/// Stream of [`Event`]s from an active subscription.
pub struct EventStream {
    lines: Lines<BufReader<OwnedReadHalf>>,
}

impl EventStream {
    /// The next event, or `None` once the server closes the connection.
    pub async fn next(&mut self) -> Option<AppletResult<Event>> {
        loop {
            match self.lines.next_line().await {
                Ok(Some(line)) if line.trim().is_empty() => continue,
                Ok(Some(line)) => return Some(Ok(parse_event(&line))),
                Ok(None) => return None,
                Err(e) => return Some(Err(e.into())),
            }
        }
    }
}

/// The wire protocol tokenizes client lines on whitespace and never
/// unescapes the command name or a field key, so an `action`/key carrying
/// whitespace would forge extra tokens or whole client lines. Values are
/// safe (they are escaped). Reject the unsafe shapes loudly.
fn validate_token(label: &str, s: &str, forbid_eq: bool) -> AppletResult<()> {
    if s.is_empty() {
        return Err(format!("IPC {label} must not be empty").into());
    }
    if s.chars().any(|c| c.is_ascii_whitespace()) {
        return Err(format!("IPC {label} {s:?} must not contain whitespace").into());
    }
    if forbid_eq && s.contains('=') {
        return Err(format!("IPC param key {s:?} must not contain '='").into());
    }
    Ok(())
}

fn escape(s: &str) -> String {
    s.replace('\\', "\\\\")
        .replace('\n', "\\n")
        .replace('\t', "\\t")
        .replace(' ', "\\s")
}

fn unescape(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut chars = s.chars();
    while let Some(c) = chars.next() {
        if c == '\\' {
            match chars.next() {
                Some('s') => out.push(' '),
                Some('n') => out.push('\n'),
                Some('t') => out.push('\t'),
                Some('\\') => out.push('\\'),
                Some(x) => {
                    out.push('\\');
                    out.push(x);
                }
                None => out.push('\\'),
            }
        } else {
            out.push(c);
        }
    }
    out
}

fn parse_event(line: &str) -> Event {
    let mut tokens = line.split_ascii_whitespace();
    let name = tokens.next().unwrap_or_default().to_owned();
    let mut ts = 0u64;
    let mut fields = HashMap::new();
    for token in tokens {
        let Some((k, v)) = token.split_once('=') else {
            continue;
        };
        let v = unescape(v);
        if k == "ts" {
            if let Ok(n) = v.parse::<u64>() {
                ts = n;
                continue;
            }
        }
        fields.insert(k.to_owned(), v);
    }
    Event { name, ts, fields }
}

fn parse_ack(line: &str) -> AppletResult<HashMap<String, String>> {
    let mut tokens = line.split_ascii_whitespace();
    if tokens.next() != Some("ack") {
        return Err(format!("expected an ack, got: {line}").into());
    }
    let mut ok = false;
    let mut fields = HashMap::new();
    for token in tokens {
        let Some((k, v)) = token.split_once('=') else {
            continue;
        };
        let v = unescape(v);
        match k {
            "ok" => ok = v == "true",
            _ => {
                fields.insert(k.to_owned(), v);
            }
        }
    }
    if ok {
        Ok(fields)
    } else {
        let msg = fields
            .remove("error")
            .unwrap_or_else(|| "command failed".to_owned());
        Err(msg.into())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::{SystemTime, UNIX_EPOCH};
    use tokio::net::UnixListener;

    fn tmp_socket() -> PathBuf {
        let n = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        std::env::temp_dir().join(format!("glimpse-ipc-test-{n}.sock"))
    }

    #[test]
    fn escape_roundtrip() {
        let s = "a b\tc\nd\\e";
        assert_eq!(unescape(&escape(s)), s);
    }

    #[test]
    fn parse_event_unescapes_and_extracts_ts() {
        let ev = parse_event("notification.received body=line1\\nline2\\sword ts=42");
        assert_eq!(ev.name, "notification.received");
        assert_eq!(ev.ts, 42);
        assert_eq!(
            ev.fields.get("body").map(String::as_str),
            Some("line1\nline2 word")
        );
    }

    #[test]
    fn validate_token_rejects_injection() {
        assert!(validate_token("action", "open_uri", false).is_ok());
        assert!(validate_token("action", "a\nsubscribe *", false).is_err());
        assert!(validate_token("action", "a b", false).is_err());
        assert!(validate_token("action", "", false).is_err());
        assert!(validate_token("param key", "k=v", true).is_err());
        assert!(validate_token("param key", "ok", true).is_ok());
    }

    #[tokio::test]
    async fn dispatch_rejects_unsafe_action_before_connecting() {
        let sub = Subscriber {
            socket: PathBuf::from("/nonexistent/glimpse-x.sock"),
        };
        let err = sub
            .dispatch("evil\naction", [("k", "v")])
            .await
            .unwrap_err();
        assert!(err.to_string().contains("whitespace"));
        // A bad key must also be rejected before any connect attempt.
        let err = sub.dispatch("ok", [("bad key", "v")]).await.unwrap_err();
        assert!(err.to_string().contains("whitespace"));
    }

    #[test]
    fn parse_ack_failure_is_err() {
        assert!(parse_ack("ack ok=false error=nope").is_err());
        let ok = parse_ack("ack ok=true echo=hi").unwrap();
        assert_eq!(ok.get("echo").map(String::as_str), Some("hi"));
    }

    #[tokio::test]
    async fn dispatch_and_listen_against_fake_server() {
        let socket = tmp_socket();
        let listener = UnixListener::bind(&socket).unwrap();
        let server = tokio::spawn(async move {
            // First connection: a dispatch.
            let (stream, _) = listener.accept().await.unwrap();
            let (r, mut w) = stream.into_split();
            w.write_all(b"hello version=test\n").await.unwrap();
            let mut lines = BufReader::new(r).lines();
            let cmd = lines.next_line().await.unwrap().unwrap();
            assert_eq!(cmd, "open_uri uri=https://example.com");
            w.write_all(b"ack ok=true echo=done\n").await.unwrap();

            // Second connection: a subscription.
            let (stream, _) = listener.accept().await.unwrap();
            let (r, mut w) = stream.into_split();
            w.write_all(b"hello version=test\n").await.unwrap();
            let mut lines = BufReader::new(r).lines();
            let sub = lines.next_line().await.unwrap().unwrap();
            assert_eq!(sub, "subscribe audio.*");
            w.write_all(b"audio.volume_changed volume=42 ts=7\n")
                .await
                .unwrap();
        });

        let sub = Subscriber { socket };
        let ack = sub
            .dispatch("open_uri", [("uri", "https://example.com")])
            .await
            .unwrap();
        assert_eq!(ack.get("echo").map(String::as_str), Some("done"));

        let mut events = sub.listen("audio.*").await.unwrap();
        let ev = events.next().await.unwrap().unwrap();
        assert_eq!(ev.name, "audio.volume_changed");
        assert_eq!(ev.ts, 7);
        assert_eq!(ev.fields.get("volume").map(String::as_str), Some("42"));

        server.await.unwrap();
    }
}
